### PR TITLE
feat(foreach): run against previously failed or successful repos

### PR DESCRIPTION
**Pull request by Dan7-7-7**
https://github.com/Skyscanner/turbolift/issues/143

### CODE REVIEW

File: cmd/foreach/foreach.go [hunk 1]
```diff
@@ -35,15 +35,19 @@ import (
...
 
 	overallResultsDirectory string
 
 	successfulResultsDirectory string
 	successfulReposFileName    string
+	successfulReposSymlink     = ".latest_successful"
 
 	failedResultsDirectory string
 	failedReposFileName    string
+	failedReposSymlink     = ".latest_failed"
```
Comment (sledigabel) [inline cmd/foreach/foreach.go:50]: those could be const?

File: cmd/foreach/foreach.go [hunk 2]
```diff
@@ -65,7 +80,9 @@ marks that no further options should be interpreted by turbolift.`,
 		Args: cobra.MinimumNArgs(1),
 	}
 
-	cmd.Flags().StringVar(&repoFile, "repos", "repos.txt", "A file containing a list of repositories to clone.")
+	cmd.Flags().StringVar(&repoFile, "repos", "", "A file containing a list of repositories to clone.")
```
Comment (sledigabel) [inline cmd/foreach/foreach.go:83]: I'm not sure this is needed - this mutex between repos.txt and successful/failed seems complicated, should we just ignore the repoFile if --successful or --failed is set?

File: cmd/foreach/foreach.go [hunk 3]
```diff
@@ -151,6 +185,27 @@ func setupOutputFiles(campaignName string, command string) {
...
 
+	if _, err := os.Lstat(successfulReposSymlink); err == nil {
+		err := os.Remove(successfulReposSymlink)
+		if err != nil {
+			logger.Warnf("Failed to remove previous symlink for successful repos: %v", err)
+		}
+	}
+	err := os.Symlink(successfulReposFileName, successfulReposSymlink)
+	if err != nil {
+		logger.Warnf("Failed to create symlink for successful repos: %v", err)
```
Comment (sledigabel) [inline cmd/foreach/foreach.go:196]: warn or fail?
Could there be a reason why this should fail?

File: cmd/foreach/foreach.go [hunk 4]
```diff
@@ -35,15 +35,19 @@ import (
...
-	repoFile = "repos.txt"
+	repoFile   string
+	successful bool
+	failed     bool
 
 	overallResultsDirectory string
 
 	successfulResultsDirectory string
 	successfulReposFileName    string
+	successfulReposSymlink     = ".latest_successful"
```
Comment (rnorth) [inline cmd/foreach/foreach.go:46]: Do we definitely need two symlinks? If we have our output directories to be something like:

```
some_temp_dir_from_os/
    some_random_bit/
        successful/
            repos.txt
            logs/
        failed/
            repos.txt
            logs/
```

Then we could just use one symlink, to the `/some_temp_dir_from_os/some_random_bit` path and infer the rest as needed.

File: cmd/foreach/foreach.go [hunk 5]
```diff
@@ -151,6 +185,27 @@ func setupOutputFiles(campaignName string, command string) {
 	defer successfulReposFile.Close()
 	defer failedReposFile.Close()
 
+	if _, err := os.Lstat(successfulReposSymlink); err == nil {
```
Comment (sledigabel) [inline cmd/foreach/foreach.go:188]: You can prob just override it?

Comment (Dan7-7-7) [reply cmd/foreach/foreach.go:188]: os.Symlink seems to error if there is a previously existing symlink with the same name

File: cmd/foreach/foreach.go (hunk 3 above)
Comment (Dan7-7-7) [reply cmd/foreach/foreach.go]: good question, I didn't think we wanted the whole thing to bomb out if this went wrong. It's not really an essential piece of functionality so perhaps better if things continue working if it breaks?

File: cmd/foreach/foreach.go (hunk 2 above)
Comment (Dan7-7-7) [reply cmd/foreach/foreach.go:83]: yeah perhaps unnecessary, I just thought that if someone is a bit confused, they could do `turbolift foreach --failed --repos repos.txt -- some command` and not realise what is going on
happy to change it back though

Comment (Dan7-7-7) [reply cmd/foreach/foreach.go]: Following discussion, I've updated this to still check that only one flag has been set, but not do the convoluted logic of resetting the default value. My example above is very unlikely and also covered by the readme instructions

File: cmd/foreach/foreach.go [hunk 6]
```diff
@@ -94,24 +94,24 @@ func runE(c *cobra.Command, args []string) error {
...
+		logger.Printf("Running against previously successful repos only")
 	} else if failed {
 		previousResults, err := os.Readlink(previousResultsSymlink)
 		if err != nil {
 			return errors.New("no previous foreach logs found")
 		}
 		repoFile = path.Join(previousResults, "failed", "repos.txt")
-	} else if !customRepoFile {
-		repoFile = "repos.txt"
+		logger.Printf("Running against previously failed repos only")
```
Comment (rnorth) [inline cmd/foreach/foreach.go:114]: Is this log line in the wrong branch?

**Review by rnorth — APPROVED** [review]
Approved in fortnightly call

**Pull request merged**